skxray/core/correlation.py

```diff
@@ -402,3 +351,366 @@ def auto_corr_scat_factor(lags, beta, relaxation_rate, baseline=1):
+
+    start_time = time.time()  # used to log the computation time (optionally)
+    #  for two time correlation
+
+    tot_channels, lag_steps = utils.multi_tau_lags(num_levels, num_bufs)
+
+    for n, img in enumerate(images):
+        cur[0] = (1 + cur[0]) % num_bufs  # increment buffer
+
+        count_level[0] = 1 + count_level[0]
+        # current image number
+        current_img_time = n + 1
+
+        # Put the image into the ring buffer.
+        buf[0, cur[0] - 1] = (np.ravel(img))[pixel_list]
+
+        # Compute the two time correlations between the first level
+        # (undownsampled) frames. two_time and img_per_level in place!
+        _two_time_process(buf, two_time, label_array, num_bufs, num_pixels,
+                          img_per_level, lag_steps, current_img_time, level=0,
+                          buf_no=cur[0] - 1)
+
+        # time frame for each level
+        time_ind[0].append(current_img_time)
+
+        # check whether the number of levels is one, otherwise
+        # continue processing the next level
+        processing = num_levels > 1
+
+        # Compute the correlations for all higher levels.
+        level = 1
+        while processing:
+            if not track_level[level]:
+                track_level[level] = 1
+                processing = False
+            else:
+                prev = 1 + (cur[level - 1] - 2) % num_bufs
+                cur[level] = 1 + cur[level] % num_bufs
+                count_level[level] = 1 + count_level[level]
+
+                buf[level, cur[level] - 1] = (buf[level - 1, prev - 1] +
+                                              buf[level - 1,
+                                                  cur[level - 1] - 1])/2
+
+                t1_idx = (count_level[level] - 1) * 2
+
+                current_img_time = ((time_ind[level - 1])[t1_idx]
+                                    + (time_ind[level - 1])[t1_idx + 1])/2.
+
+                # time frame for each level
+                time_ind[level].append(current_img_time)
+
+                # make the track_level zero once that level is processed
+                track_level[level] = 0
+
+                # call the _two_time_process function for each multi-tau level
+                # for multi-tau levels greater than one
+                # Again, this is modifying things in place. See comment
+                # on previous call above.
+                _two_time_process(buf, two_time, label_array, num_bufs,
+                                  num_pixels, img_per_level, lag_steps,
+                                  current_img_time, level=level,
+                                  buf_no=cur[level]-1)
+                level += 1
+
+                # Checking whether there is next level for processing
+                processing = level < num_levels
+
+    for q in range(np.max(labels)):
+        x0 = two_time[:, :, q]
+        two_time[:, :, q] = (np.tril(x0) + np.tril(x0).T
```

Comment:
Is there data in the upper triangle that is being over-written here?

It is probably better to return the un-patched up result and provide a function to do this as needed.

If there is not data in the upper triangle would `x0 + np.tril(x0, -1)` would be faster.